### ISSUE TITLE
Email address not visible under user account settings

### ISSUE THREAD

**Issue by larciere** [vector-im/element-web#15062]
### Description

Email address not visible in user account settings

### Steps to reproduce

- Add email address (settings > general > profile > email address)
- Confirm email address 
- Now it shows the email address
- Login again
- No email address visible

I would expect to see my email address and also be able to change this if needed.

### Version information

- **Platform**: Web
- **Version**: Element-versie: 1.7.4

**Comment by t3chguy** [comment]
I cannot reproduce this using the matrix.org Synapse

After you confirm the email address did you go back to the Element where you added it, clicked "Continue" and provided your password if you were asked to finish adding the address?

What server were you using?

**Comment by plankobostjan** [comment]
I'm experiencing the same issue as @larciere.

I did the following:

1. added the email address under _Setting > General > Email addresses_
2. confirmed it via the link I got in the email
3. returned back to the Element and clicked _Continue_
4. my email was shown under _Setting > General > Email addresses_ with a _Delete_ button beside it
5. I've closed the settings menu

Now, if I reopen the settings menu, no email address is shown under _Email addresses_. If I try to re-add my email, I get an error saying that it's already in use.

I was able to reproduce this on Web and Desktop client.

### Version information
Platform: Web
Element version: 1.7.12

Platform: Desktop
Element version: 1.7.13

I'm running my own Synapse server (matrix-synapse-1.23.0).

**Comment by t3chguy** [comment]
> I'm running my own Synapse server (matrix-synapse-1.23.0).

This is likely at fault, I'd wager its worth reporting the bug there, what combination of settings around emails are you using in Synapse?

**Comment by plankobostjan** [comment]
Yes, I agree. At first I've thought that the issue is with Element, now I'm pretty sure that the problem is Synapse. Though before opening an issue there, I want to do a bit more debugging myself. 

As far as email settings go, I'm using external server for sending emails. Port 587, with STARTTLS enabled by default. And I'm using username and password for authentication with the mail server. I also have _enable_notifs_ set to true. Other settings regarding email are left as default.

**Comment by plankobostjan** [comment]
I did some debugging of this issue today, and I've noticed something rather strange.

Since I was unable to find anything useful in _homeserver.log_, I've decided to take a look at the requests made to my homeserver. Here's what I've found.

When I open the settings dialog, there's - among bunch of other requests - this:
```
REQUEST:
OPTIONS /_matrix/client/r0/account/3pid HTTP/1.1
Host: matrix.myserver.com
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0
Accept: */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate, br
Access-Control-Request-Method: GET
Access-Control-Request-Headers: authorization
Origin: https://riot.myserver.com
Connection: keep-alive
DNT: 1
Sec-GPC: 1

RESPONSE:
HTTP/1.1 204 No Content
Server: nginx
Date: Sat, 28 Nov 2020 06:40:14 GMT
Content-Length: 0
Connection: keep-alive
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: GET, HEAD, POST, PUT, DELETE, OPTIONS
Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization
X-Frame-Options: SAMEORIGIN
Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
X-XSS-Protection: 1; mode=block
```
And following the above there's this:
```
REQUEST:
GET /_matrix/client/r0/account/3pid HTTP/1.1
Host: matrix.myserver.com
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0
Accept: application/json
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate, br
Origin: https://riot.myserver.com
DNT: 1
Sec-GPC: 1
Authorization: Bearer my_access_token
Connection: keep-alive

RESPONSE:
HTTP/1.1 200 OK
Server: nginx
Date: Sat, 28 Nov 2020 06:40:14 GMT
Content-Type: application/json
Transfer-Encoding: chunked
Connection: keep-alive
Cache-Control: no-cache, no-store, must-revalidate
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: GET, HEAD, POST, PUT, DELETE, OPTIONS
Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization
X-Frame-Options: SAMEORIGIN
Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
X-XSS-Protection: 1; mode=block

{"threepids":[{"medium":"email","address":"myusername@myemail.com","validated_at":12345678901,"added_at":1234567890112}]}
```
Also, I see this in _homeserver.log_ everytime I open the settings dialog:
```
2020-11-28 10:15:13,188 - synapse.access.http.8008 - 331 - INFO - GET-1042 - - - 8008 - {@user:myserver.com} Processed request: 0.004sec/-0.000sec (0.002sec, 0.000sec) (0.001sec/0.001sec/1) 126B 200 "GET /_matrix/client/r0/account/3pid HTTP/1.0" "Mozilla/5.0 (X11; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0" [0 dbevts]
```
It is strange though, because in Element, the element logo spins for a few seconds, as if processing something. And then no email address is shown.

But it works on matrix.org homeserver (I have an account there too). I've logged in (to matrix.org) via my Element client, and 3pids were shown as expected.

I hope this is helpful. Never did this kind of debugging before... 

I'm confused now.. Is this a problem with Element or Synapse? Any ideas what or where might be the problem?

**Comment by jryans** [comment]
Hmm, that is surprising... 😕 

What does your server report via `GET /_matrix/client/versions`? For example, [here](https://matrix-client.matrix.org/_matrix/client/versions) is matrix.org:

<img width="409" alt="image" src="https://user-images.githubusercontent.com/279572/100870007-d13adf80-3495-11eb-8f91-2d5246fd85b5.png">

**Comment by plankobostjan** [comment]
Hmmm... Result of `GET /_matrix/client/versions` on my server is exactly the same as on matrix.org:

![img-2020-12-03-063810](https://user-images.githubusercontent.com/17665959/100968847-6ab7cf00-3532-11eb-8841-5119bf0c0b15.png)

**Comment by jryans** [comment]
Hmm, what about in your Element `config.json`?

**Comment by plankobostjan** [comment]
Here you go:
```
{
    "default_server_config": {
        "m.homeserver": {
            "base_url": "https://matrix.myserver.com",
            "server_name": "myserver.com"
        },
        "m.identity_server": {
            "base_url": "https://vector.im"
        }
    },
    "disable_custom_urls": true,
    "disable_guests": true,
    "disable_login_language_selector": false,
    "disable_3pid_login": true,
    "brand": "Element",
    "integrations_ui_url": "https://scalar.vector.im/",
    "integrations_rest_url": "https://scalar.vector.im/api",
    "integrations_widgets_urls": [
        "https://scalar.vector.im/_matrix/integrations/v1",
        "https://scalar.vector.im/api",
        "https://scalar-staging.vector.im/_matrix/integrations/v1",
        "https://scalar-staging.vector.im/api",
        "https://scalar-staging.riot.im/scalar/api"
    ],
    "bug_report_endpoint_url": "https://element.io/bugreports/submit",
    "defaultCountryCode": "SI",
    "showLabsSettings": true,
    "features": {
        "feature_new_spinner": true
    },
    "default_federate": false,
    "default_theme": "light",
    "roomDirectory": {
        "servers": [
            "myserver.com"
        ]
    },
    "piwik": "false",
    "enable_presence_by_hs_url": {
        "https://matrix.org": false,
        "https://matrix-client.matrix.org": false
    },
    "settingDefaults": {
        "breadcrumbs": true
    },
    "jitsi": {
        "preferredDomain": "jitsi.riot.im"
    },
    "embeddedPages": {
    "welcomeUrl": "welcome.html"
    },
    "permalinkPrefix": "https://riot.myserver.com",
    "terms_and_conditions_links": [
        {
            "url": "https://element.io/privacy",
            "text": "Privacy Policy"
        },
        {
            "url": "https://element.io/cookie-policy",
            "text": "Cookie Policy"
        }
    ]
}
```
I've also tried changing some settings to see if it would make a difference, but so far, no luck.

**Comment by jryans** [comment]
Hmm, I don't see anything there that would overlap with this issue... 😕 

Are there any network requests failing? Maybe if you submit debug logs after having opened Settings would help spot something.

**Comment by plankobostjan** [comment]
Yes, there are requests failing. Per your suggestion I took a look at debug logs, and found this:

```
log-11.log:2020-12-12T07:52:43.544Z W request failed: CORS request rejected: https://matrix.myserver.com/_matrix/client/r0/account/3pid
log-12.log:2020-12-12T07:52:31.283Z W request failed: CORS request rejected: https://matrix.myserver.com/_matrix/client/r0/account/3pid
log-28.log:2020-11-28T07:14:30.088Z W request failed: CORS request rejected: https://matrix.myserver.com/_matrix/client/r0/account/3pid
log-28.log:2020-11-28T07:14:32.416Z W request failed: CORS request rejected: https://matrix.myserver.com/_matrix/client/r0/account/3pid

```

Now, is that a problem with my nginx (reverse proxy) configuration, or is this a problem with Element?

Just to note. 3pids work fine on Element for Android.

**Comment by jryans** [comment]
It seems like those requests might be missing some CORS-related headers then... Do the failing requests have `Access-Control-Allow-Origin` headers?

**Comment by plankobostjan** [comment]
WOW! Strange. This is now resolved, I guess, though I don't know why. Here's what I've done.

I went to the browser again, to check failing requests. There were none. At lease not for 3PIDs. But there were couple of "500 Internal Server Error" messages about reaching vector.im identity server. That seemed strange. Though I noticed Element complaining about being unable to reach vector.im before, I never paid much attention to these, since I'm not using an identity server. Anyway. I then clicked "Disconnect" under "Identity Server". After a while I got a message that identity server is unreachable. I clicked "Disconnect anyway". And viola, my 3PIDs showed up immediately. They now work on all clients.

As far as I'm concerned, you can close this issue. Though I am really curious what was the cause of all this. I don't know. Maybe vector.im being unreachable somehow caused 3PIDs not showing up?

**Comment by jryans** [comment]
Hmm, very interesting! I am glad to hear the core issue is resolved for you at least!

Perhaps there is some bug with handling 3PID responses when the IS is unavailable, so let's keep this issue to track that for investigation.

**Comment by MadLittleMods** [comment]
## CORS headers

> It seems like those requests might be missing some CORS-related headers then... Do the failing requests have `Access-Control-Allow-Origin` headers?

Yes, the `/_matrix/client/r0/account/3pid` requests have CORS headers so they can be accessed anywhere.

`OPTIONS https://matrix-client.matrix.org/_matrix/client/r0/account/3pid`

```
access-control-allow-headers: Origin, X-Requested-With, Content-Type, Accept, Authorization
access-control-allow-methods: GET, POST, PUT, DELETE, OPTIONS
access-control-allow-origin: *
```

`GET https://matrix-client.matrix.org/_matrix/client/r0/account/3pid`

```
access-control-allow-headers: X-Requested-With, Content-Type, Authorization, Date
access-control-allow-methods: GET, HEAD, POST, PUT, DELETE, OPTIONS
access-control-allow-origin: *
```


## Needs better error handling

> Perhaps there is some bug with handling 3PID responses when the IS is unavailable, so let's keep this issue to track that for investigation.

If you use the network devtools to block the `https://matrix-client.matrix.org/_matrix/client/r0/account/3pid` request and open the user settings modal, the email/phone number list is just empty with no indication of an error.

![](https://user-images.githubusercontent.com/558581/132068114-70e42065-3e06-44f7-8a0e-6e0f062cea36.png)

There is a `console.warn` in the devtools console from `rageshake.js`
```
ConnectionError: request failed: CORS request rejected: https://matrix-client.matrix.org/_matrix/client/r0/account/3pid
    at eval (http-api.js?5b4b:801)
    at Object.eval [as callback] (http-api.js?5b4b:742)
    at on_response (index.js?03e1:266)
    at XMLHttpRequest.on_state_change (index.js?03e1:242)
```
